File: python/src/payskill/keychain.py
"""OS keychain reader -- reads the private key stored by `pay` CLI.

Service: "pay", account: "default".

keyring is an optional dependency. If not installed, returns None.
"""

from __future__ import annotations


def read_from_keychain() -> str | None:
    """Read private key from OS keychain.

    Returns the key as a hex string, or None if keyring is not
    installed or no key is stored.
    """
    try:
        import keyring  # type: ignore[import-untyped]

        return keyring.get_password("pay", "default")  # type: ignore[no-any-return]
    except Exception:
        return None

File: typescript/src/keychain.ts
/**
 * OS keychain reader — reads the private key stored by `pay` CLI.
 * Service: "pay", account: "default".
 *
 * keytar is an optional dependency. If not installed, returns null.
 */

export async function readFromKeychain(): Promise<string | null> {
  try {
    const moduleName = "keytar";
    const keytar = (await import(moduleName)) as {
      default: { getPassword(service: string, account: string): Promise<string | null> };
    };
    return await keytar.default.getPassword("pay", "default");
  } catch {
    return null;
  }
}
